ext/semian_cb_data/semian_cb_data.c

```diff
@@ -0,0 +1,900 @@
+  flags = IPC_EXCL | IPC_CREAT | permissions;
+
+  semid = semget(key, kCBSemaphoreCount, flags);
+  if (semid >= 0) {
+    *created = 1;
+    //rb_warn("semget: received %d semaphore(s) with key %d, semid %d", kCBSemaphoreCount, key, semid);
+  } else if (semid == -1 && errno == EEXIST) {
+    flags &= ~IPC_EXCL;
+    semid = semget(key, kCBSemaphoreCount, flags);
+    //rb_warn("semget: retrieved existing semaphore with key %d, semid %d", key, semid);
+  }
+  return semid;
+}
+
+
+
+
+/*
+ * Create or acquire previously made semaphore
+ */
+
+static VALUE
+semian_cb_data_acquire_semaphore (VALUE self, VALUE permissions)
+{
+  // Function flow, semaphore creation methods are
+  //   borrowed from semian.c since they have been previously tested
+
+  semian_cb_data *ptr;
+  TypedData_Get_Struct(self, semian_cb_data, &semian_cb_data_type, ptr);
+
+  if (TYPE(permissions) != T_FIXNUM)
+    rb_raise(rb_eTypeError, "expected integer for permissions");
+
+  // bool for initializing (configure_tickets) or not
+  int created = 0;
+  key_t key = ptr->key;
+  int semid = create_semaphore(key, FIX2LONG(permissions), &created);
+  if (-1 == semid) {
+    raise_semian_syscall_error("semget()", errno);
+  }
+  ptr->semid = semid;
+
+  // initialize to 1 and set permissions
+  configure_tickets(ptr->semid, kCBTicketMax, created);
+  set_semaphore_permissions(ptr->semid, FIX2LONG(permissions));
+
+  return self;
+}
+
+
+static VALUE
+semian_cb_data_delete_semaphore(VALUE self)
+{
+  semian_cb_data *ptr;
+  TypedData_Get_Struct(self, semian_cb_data, &semian_cb_data_type, ptr);
+  if (-1 == ptr->semid) // do nothing if semaphore not acquired
+    return self;
+
+  if (-1 == semctl(ptr->semid, 0, IPC_RMID)) {
+    if (EIDRM == errno) {
+      rb_warn("semctl: failed to delete semaphore set with semid %d: already removed", ptr->semid);
+      ptr->semid = -1;
+    } else {
+      rb_warn("semctl: failed to remove semaphore with semid %d, errno %d (%s)",ptr->semid, errno, strerror(errno));
+    }
+  } else {
+    //rb_warn("semctl: semaphore set with semid %d deleted", ptr->semid);
+    ptr->semid = -1;
+  }
+  return self;
+}
+
+
+
+
+/*
+ * semian_cb_data_lock/unlock and associated functions decrement/increment semaphore
+ */
+
+static VALUE
+semian_cb_data_lock(VALUE self)
+{
+  return (VALUE) WITHOUT_GVL(semian_cb_data_lock_without_gvl, (void *)self, RUBY_UBF_IO, NULL);
+}
+
+static void *
+semian_cb_data_lock_without_gvl(void *self)
+{
+  semian_cb_data *ptr;
+  TypedData_Get_Struct((VALUE)self, semian_cb_data, &semian_cb_data_type, ptr);
+  if (ptr->lock_triggered)
+    return (void *)Qtrue;
+  if (-1 == ptr->semid){
+    rb_raise(eInternal, "semid not set, errno %d: (%s)", errno, strerror(errno));
+    return (void *)Qfalse;
+  }
+  VALUE retval;
+
+  struct timespec ts = { 0 };
+  ts.tv_sec = kCBInternalTimeout;
+
+  if (-1 == semtimedop(ptr->semid,&decrement,1, &ts)) {
+    rb_raise(eInternal, "error with semop locking,  %d: (%s)", errno, strerror(errno));
```

Comment:
Error could be a bit more desciptive: `Error acquiring semaphore to mutate circuit breaker structure: %d (%s)`